pkg/storage/tsdb/bucketindex/loader.go

```diff
@@ -0,0 +1,269 @@
+			Help: "Total number of bucket index loading failures.",
+		}),
+		loadDuration: promauto.With(reg).NewHistogram(prometheus.HistogramOpts{
+			Name:    "cortex_bucket_index_load_duration_seconds",
+			Help:    "Duration of the a single bucket index loading operation in seconds.",
+			Buckets: []float64{0.01, 0.02, 0.05, 0.1, 0.2, 0.3, 1, 10},
+		}),
+	}
+
+	l.loaded = promauto.With(reg).NewGaugeFunc(prometheus.GaugeOpts{
+		Name: "cortex_bucket_index_loaded",
+		Help: "Number of bucket indexes currently loaded in-memory.",
+	}, l.countLoadedIndexesMetric)
+
+	// Apply a jitter to the sync frequency in order to increase the probability
+	// of hitting the shared cache (if any).
+	checkInterval := util.DurationWithJitter(cfg.CheckInterval, 0.2)
+	l.Service = services.NewTimerService(checkInterval, nil, l.checkCachedIndexes, nil)
+
+	return l
+}
+
+// GetIndex returns the bucket index for the given user. It returns the in-memory cached
+// index if available, or load it from the bucket otherwise.
+func (l *Loader) GetIndex(ctx context.Context, userID string) (*Index, error) {
+	l.indexesMx.RLock()
+	if entry := l.indexes[userID]; entry != nil {
+		idx := entry.index
+		err := entry.err
+		l.indexesMx.RUnlock()
+
+		// We don't check if the index is stale because it's the responsibility
+		// of the background job to keep it updated.
+		entry.requestedAt.Store(time.Now().Unix())
+		return idx, err
+	}
+	l.indexesMx.RUnlock()
+
+	startTime := time.Now()
+	l.loadAttempts.Inc()
+	idx, err := ReadIndex(ctx, l.bkt, userID, l.logger)
+	if err != nil {
+		// Cache the error, to avoid hammering the object store in case of persistent issues
+		// (eg. corrupted bucket index or not existing).
+		l.cacheIndex(userID, nil, err)
+
+		l.loadFailures.Inc()
+		if !errors.Is(err, ErrIndexNotFound) {
+			level.Error(l.logger).Log("msg", "unable to load bucket index", "user", userID, "err", err)
+		}
+
+		return nil, err
+	}
+
+	// Cache the index.
+	l.cacheIndex(userID, idx, nil)
+
+	elapsedTime := time.Since(startTime)
+	l.loadDuration.Observe(elapsedTime.Seconds())
+	level.Info(l.logger).Log("msg", "loaded bucket index", "user", userID, "duration", elapsedTime)
+	return idx, nil
+}
+
+func (l *Loader) cacheIndex(userID string, idx *Index, err error) {
+	l.indexesMx.Lock()
+	defer l.indexesMx.Unlock()
+
+	// Not an issue if, due to concurrency, another index was already cached
+	// and we overwrite it: last will win.
+	l.indexes[userID] = newCachedIndex(idx, err)
+}
+
+// checkCachedIndexes checks all cached indexes and, for each of them, does two things:
+// 1. Offload indexes not requested since >= idle timeout
+// 2. Update indexes which have been updated last time since >= update timeout
+func (l *Loader) checkCachedIndexes(ctx context.Context) error {
+	var (
+		toUpdate []string
+		toDelete []string
+		now      = time.Now()
+	)
+
+	// Build a list of users for which we should update or delete the index.
+	l.indexesMx.RLock()
```

Comment:
I would suggest making this a helper method to calculate toDelete and toUpdate since it's the scope that needs the lock and doing something separate to make things slightly easier to read.

Reply:
Good point, done. What do you think=

Reply:
Thanks, LGTM